### PR TITLE
condense and consolidate types/actions with generic

### PR DESCRIPTION
**Pull request by EDsCODE**
-use generics to reduce code needed to create actions and infer type
-improvement is to reduce amount of declarations needed in the types file (e.g. declaring the action shape with payloads when it can be inferred) also reduces amount of imports needed

### CODE REVIEW

File: app/client/reduxUtils.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  }
+}
+
+type FunctionType = (...args: any[]) => any;
+type ActionCreatorsMapObject = { [actionCreator: string]: FunctionType };
+
+export type ActionsUnion<T extends ActionCreatorsMapObject> = ReturnType<
+  T[keyof T]
+>;
+
```
Comment (EDsCODE) [inline app/client/reduxUtils.ts]: Reference: https://medium.com/@martin_hotell/improved-redux-type-safety-with-typescript-2-8-2c11a8062575

**Comment by EDsCODE** [comment]
**Review for new reducer format only not merging purposes. Once frontend-testing is merged I will create tests to make sure all the changes I made in this PR are solid

**Comment by EDsCODE** [comment]
@tristanhcole kinda dense but when you get the chance to look through and confirmed, I can start refactoring all the redux to be like this!

File: app/client/reducers/auth/types.ts [hunk 2]
```diff
@@ -45,128 +40,60 @@ export interface LoginSuccess {
...
-interface ResetPasswordFailure {
-  type: typeof RESET_PASSWORD_FAILURE;
-  error: string;
-}
-export type ResetPasswordAction =
-  | ResetPasswordRequest
-  | ResetPasswordSuccess
-  | ResetPasswordFailure;
 
 export interface AdminUser {
```
Comment (tristanhcole) [inline app/client/reducers/auth/types.ts]: NIT: Given all other interfaces for payloads include "Payload" in variable name, this should follow the same pattern... i.e. `AdminUserPayload`

File: app/client/reducers/auth/types.ts [hunk 3]
```diff
@@ -177,49 +104,31 @@ export interface AdminUser {
   is_disabled: boolean;
 }
 export const UPDATE_ADMIN_USER_LIST = "UPDATE_ADMIN_USER_LIST";
-interface UpdateAdminUsers {
-  type: typeof UPDATE_ADMIN_USER_LIST;
-  admins: AdminUser;
-}
-export type AdminUserListAction = UpdateAdminUsers;
 
 export interface Invite {
```
Comment (tristanhcole) [inline app/client/reducers/auth/types.ts]: `InvitePayload`

File: app/client/reducers/auth/types.ts (hunk 2 above)
Comment (tristanhcole) [reply app/client/reducers/auth/types.ts]: Pretty sure I wrote this part of the code lol

File: app/client/sagas/authSagas.js [hunk 4]
```diff
@@ -1,11 +1,8 @@
 import {
```
Comment (tristanhcole) [inline app/client/sagas/authSagas.js]: Definitely think we should be also converting sagas to typescript

File: app/client/sagas/authSagas.js [hunk 5]
```diff
@@ -453,33 +471,36 @@ function* inviteUserRequest({ payload }) {
...
     const validateTFAresponse = yield call(ValidateTFAAPI, payload);
+
+    yield put(ValidateTfaAction.validateTFASuccess());
     yield call(updateOrganisationStateFromLoginData, validateTFAresponse);
-    yield put({ type: VALIDATE_TFA_SUCCESS, validateTFAresponse });
     yield call(storeTFAToken, validateTFAresponse.tfa_auth_token);
-    yield put(createLoginSuccessObject(validateTFAresponse));
+    yield put(
+      LoginAction.loginSuccess(createLoginSuccessObject(validateTFAresponse))
+    );
```
Comment (tristanhcole) [inline app/client/sagas/authSagas.js:491]: resolved merge conflict

File: app/client/reducers/auth/types.ts (hunk 2 above)
Comment (tristanhcole) [reply app/client/reducers/auth/types.ts:98]: actually just realized this doesn't make sense. scratch that

File: app/client/sagas/authSagas.js (hunk 4 above)
Comment (EDsCODE) [reply app/client/sagas/authSagas.js:1]: yeah for sure. i was going to do that in this PR too but it was getting chaotic and I'm working on keeping the changes more incremental

**Comment by EDsCODE** [comment]
> Sorry for the long wait on this - some pretty big structural changes in here and it took me some time to fully reason through all of them. So with that said, I think it's great and ready to go!
> 
> Side note: I _love_ typed payloads for actions - it really covers my biggest dislike of the payload pattern!

Awesome! Thanks guys. Before I merge this, I want to setup some tests to keep coverage on the redux. Going to make a branch off this with tests and once that's done we can merge that branch into this one and into master.

**Comment by tristanhcole** [comment]
> > Sorry for the long wait on this - some pretty big structural changes in here and it took me some time to fully reason through all of them. So with that said, I think it's great and ready to go!
> > Side note: I _love_ typed payloads for actions - it really covers my biggest dislike of the payload pattern!
> 
> Awesome! Thanks guys. Before I merge this, I want to setup some tests to keep coverage on the redux. Going to make a branch off this with tests and once that's done we can merge that branch into this one and into master.

IMO I think we could just merge this in and then write tests in another PR. Otherwise, it will turn into a behemoth/get out of date.

**Pull request merged**